addons/crm/tests/test_crm_lead.py

```diff
@@ -689,3 +689,21 @@ def test_no_date_closed_update_between_won_stages(self):
             leads.stage_id = self.stage_team1_won
         self.assertEqual(won_lead.date_closed, date_closed)
         self.assertEqual(self.lead_1.date_closed, datetime.strptime('2020-02-02 18:00', '%Y-%m-%d %H:%M'))
+
+    @users('admin')
+    def test_crm_lead_company_currency(self):
+        company = self.env['res.company'].create({
+            'name': 'Company A',
+            'currency_id': self.env.ref('base.USD').id,
+            'country_id' : self.env.ref('base.us').id,
+        })
+        lead = self.env['crm.lead'].create({
+            'name': 'Lead 1',
+            'company_id': company.id
+        })
+        self.assertEqual(lead.company_currency, self.env.ref('base.USD'))
+
+        company.currency_id = self.env.ref('base.CHF')
+        self.env.company = company
```

Comment:
You should probably use with_company instead of tweaking the low level environment, better to be aligned with common code